junit5/src/main/java/org/jboss/weld/junit5/WeldJunit5Extension.java

```diff
@@ -0,0 +1,143 @@
+/*
+ * JBoss, Home of Professional Open Source
+ * Copyright 2017, Red Hat, Inc., and individual contributors
+ * by the @authors tag. See the copyright.txt in the distribution for a
+ * full listing of individual contributors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jboss.weld.junit5;
+
+import java.lang.annotation.Annotation;
+import java.lang.reflect.Field;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+import javax.inject.Qualifier;
+
+import org.jboss.weld.config.ConfigurationKey;
+import org.jboss.weld.environment.se.Weld;
+import org.jboss.weld.environment.se.WeldContainer;
+import org.jboss.weld.junit5.annotations.WeldSetup;
+import org.junit.jupiter.api.TestInstance;
+import org.junit.jupiter.api.extension.AfterAllCallback;
+import org.junit.jupiter.api.extension.AfterTestExecutionCallback;
+import org.junit.jupiter.api.extension.ExtensionContext;
+import org.junit.jupiter.api.extension.ParameterContext;
+import org.junit.jupiter.api.extension.ParameterResolutionException;
+import org.junit.jupiter.api.extension.ParameterResolver;
+import org.junit.jupiter.api.extension.TestInstancePostProcessor;
+
+/**
+ * JUnit 5 extension allowing to bootstrap Weld SE container for each @Test method and tear it down afterwards. Also allows to
+ * inject CDI beans as parameters to @Test methods and resolves all @Inject fields in test class.
+ *
+ * @author <a href="mailto:manovotn@redhat.com">Matej Novotny</a>
+ */
+public class WeldJunit5Extension implements AfterAllCallback, TestInstancePostProcessor, AfterTestExecutionCallback, ParameterResolver {
+
+    WeldInitiator initiator;
+    WeldContainer container;
+
+    @Override
+    public void afterAll(ExtensionContext context) throws Exception {
+        if (determineTestLifecycle(context).equals(TestInstance.Lifecycle.PER_CLASS)) {
+            initiator.shutdownWeld();
+        }
+    }
+
+    @Override
+    public void afterTestExecution(ExtensionContext context) throws Exception {
+        if (determineTestLifecycle(context).equals(TestInstance.Lifecycle.PER_METHOD)) {
+            initiator.shutdownWeld();
+        }
+    }
+
+    @Override
+    public void postProcessTestInstance(Object testInstance, ExtensionContext context) throws Exception {
+
+        // obtain WeldInitiator if defined, we have to use reflections here
```

Comment:
We still need to use reflection here to obtain the actual `WeldInitiator` instance user configured. Not sure this would work if there is a class hierarchy of tests, will have to try.

Reply:
This also fails if the user declares the field private, we would need to be more rude there.